File: 2024/day 5/AOC5.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Rule struct {
	X, Y int
}

func parseInputFile(filePath string) ([]Rule, [][]int, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return nil, nil, err
	}

	// Handle both Unix and Windows line endings
	contentStr := strings.ReplaceAll(string(content), "\r\n", "\n")
	sections := strings.Split(strings.TrimSpace(contentStr), "\n\n")
	if len(sections) != 2 {
		return nil, nil, fmt.Errorf("invalid input format: expected 2 sections, got %d", len(sections))
	}

	// Parse rules
	var rules []Rule
	for _, line := range strings.Split(strings.TrimSpace(sections[0]), "\n") {
		parts := strings.Split(line, "|")
		if len(parts) == 2 {
			x, _ := strconv.Atoi(parts[0])
			y, _ := strconv.Atoi(parts[1])
			rules = append(rules, Rule{X: x, Y: y})
		}
	}

	// Parse updates
	var updates [][]int
	for _, line := range strings.Split(strings.TrimSpace(sections[1]), "\n") {
		parts := strings.Split(line, ",")
		var update []int
		for _, p := range parts {
			num, _ := strconv.Atoi(p)
			update = append(update, num)
		}
		updates = append(updates, update)
	}

	return rules, updates, nil
}

func buildGraph(rules []Rule, pages map[int]bool) (map[int][]int, map[int]int) {
	graph := make(map[int][]int)
	inDegree := make(map[int]int)

	for _, rule := range rules {
		if pages[rule.X] && pages[rule.Y] {
			graph[rule.X] = append(graph[rule.X], rule.Y)
			inDegree[rule.Y]++
			if _, exists := inDegree[rule.X]; !exists {
				inDegree[rule.X] = 0
			}
		}
	}

	return graph, inDegree
}

func topologicalSort(update []int, graph map[int][]int, inDegree map[int]int) []int {
	inDegreeCopy := make(map[int]int)
	for k, v := range inDegree {
		inDegreeCopy[k] = v
	}

	var queue []int
	for _, node := range update {
		if inDegreeCopy[node] == 0 {
			queue = append(queue, node)
		}
	}

	var sortedOrder []int
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		sortedOrder = append(sortedOrder, node)

		for _, neighbor := range graph[node] {
			inDegreeCopy[neighbor]--
			if inDegreeCopy[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	return sortedOrder
}

func isValidOrder(update []int, graph map[int][]int, inDegree map[int]int) bool {
	sorted := topologicalSort(update, graph, inDegree)
	if len(sorted) != len(update) {
		return false
	}
	for i := range update {
		if update[i] != sorted[i] {
			return false
		}
	}
	return true
}

func findMiddle(pageList []int) int {
	return pageList[len(pageList)/2]
}

func main() {
	filePath := "rules.txt"
	rules, updates, err := parseInputFile(filePath)
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return
	}

	totalMiddleSumValid := 0
	totalMiddleSumCorrected := 0

	for _, update := range updates {
		pages := make(map[int]bool)
		for _, page := range update {
			pages[page] = true
		}

		graph, inDegree := buildGraph(rules, pages)

		if isValidOrder(update, graph, inDegree) {
			totalMiddleSumValid += findMiddle(update)
		} else {
			correctedUpdate := topologicalSort(update, graph, inDegree)
			totalMiddleSumCorrected += findMiddle(correctedUpdate)
		}
	}

	fmt.Printf("Sum of middle pages from correctly ordered updates: %d\n", totalMiddleSumValid)
	fmt.Printf("Sum of middle pages after correcting invalid updates: %d\n", totalMiddleSumCorrected)
}

File: 2024/day 5/AOC5.py
from collections import defaultdict, deque

def parse_input_file(file_path):
    """Parse input from a file containing rules and updates."""
    with open(file_path, "r") as file:
        input_data = file.read()

    rules_section, updates_section = input_data.strip().split("\n\n")

    # Parse ordering rules
    rules = []
    for line in rules_section.strip().split("\n"):
        X, Y = map(int, line.split("|"))
        rules.append((X, Y))

    # Parse updates
    updates = []
    for line in updates_section.strip().split("\n"):
        updates.append(list(map(int, line.split(","))))

    return rules, updates

def build_graph(rules, pages):
    """Build a directed graph from the rules relevant to the given pages."""
    graph = defaultdict(list)
    in_degree = defaultdict(int)

    # Add edges only for the relevant pages
    for X, Y in rules:
        if X in pages and Y in pages:
            graph[X].append(Y)
            in_degree[Y] += 1
            if X not in in_degree:
                in_degree[X] = 0

    return graph, in_degree

def is_valid_order(update, graph, in_degree):
    """Check if the given update is a valid topological order."""
    # Perform topological sort using Kahn's algorithm
    in_degree_copy = in_degree.copy()
    queue = deque([node for node in update if in_degree_copy[node] == 0])
    sorted_order = []

    while queue:
        node = queue.popleft()
        sorted_order.append(node)

        for neighbor in graph[node]:
            in_degree_copy[neighbor] -= 1
            if in_degree_copy[neighbor] == 0:
                queue.append(neighbor)

    # Check if sorted_order matches the update sequence
    return sorted_order == update

def correct_order(update, graph, in_degree):
    """Generate the correct order of pages based on topological sorting."""
    in_degree_copy = in_degree.copy()
    queue = deque([node for node in update if in_degree_copy[node] == 0])
    sorted_order = []

    while queue:
        node = queue.popleft()
        sorted_order.append(node)

        for neighbor in graph[node]:
            in_degree_copy[neighbor] -= 1
            if in_degree_copy[neighbor] == 0:
                queue.append(neighbor)

    return sorted_order

def find_middle(page_list):
    """Find the middle page in a list of pages."""
    return page_list[len(page_list) // 2]

def solve_puzzle_from_file(file_path):
    """Solve the puzzle using input from a file."""
    rules, updates = parse_input_file(file_path)
    total_middle_sum_valid = 0
    total_middle_sum_corrected = 0

    for update in updates:
        # Build a graph for the current update
        graph, in_degree = build_graph(rules, set(update))

        # Check if the update is valid
        if is_valid_order(update, graph, in_degree):
            # Add the middle page to the total sum of valid updates
            total_middle_sum_valid += find_middle(update)
        else:
            # Correct the order and add the middle page to the total sum of corrected updates
            corrected_update = correct_order(update, graph, in_degree)
            total_middle_sum_corrected += find_middle(corrected_update)

    return total_middle_sum_valid, total_middle_sum_corrected

# Example usage
file_path = "rules.txt"
valid_sum, corrected_sum = solve_puzzle_from_file(file_path)
print("Sum of middle pages from correctly ordered updates:", valid_sum)
print("Sum of middle pages after correcting invalid updates:", corrected_sum)
